counsel.el

```diff
@@ -124,7 +124,9 @@ To skip the `executable-find' check, start the string with a space."
                      (car (split-string cmd)))))
       (or (and (stringp program)
                (not (string= program ""))
-               (executable-find program))
+               (executable-find program
+                                (and (version<= "27.1" emacs-version)
+                                     (tramp-tramp-file-p default-directory))))
```

Comment:
This won't work because, prior to Emacs 27, `executable-find` accepted only a single argument.

Reply:
True. I pushed it like this before seeing your suggestion, do you have a preference? https://github.com/abo-abo/swiper/pull/2618/commits/7314227f38c7a9619957520bf799c0082cb11869

Reply:
> do you have a preference?

No, it's fine either way. My only suggestion would be to write `(< emacs-major-version 27)` instead.

Reply:
Done: https://github.com/abo-abo/swiper/pull/2618/commits/7630a885613ac86f7b508792b4ae5372d870069b

Reply:
> Done: 7630a88

Not quite: you're still using version string comparison against `emacs-version`, whereas I suggested numerical comparison against `emacs-major-version`.

Reply:
Aw derp: https://github.com/abo-abo/swiper/pull/2618/commits/7c73f5b3b0518d80d8974de40de03abbd3c9c4a1

Reply:
Nearly there! The current condition excludes Emacs 27, though. It should either be:

```el
(if (<= 27 emacs-major-version)
    (executable-find program (tramp-tramp-file-p default-directory))
  (executable-find program))
```

or, as I suggested earlier:

```el
(if (< emacs-major-version 27)
    (executable-find program)
  (executable-find program (tramp-tramp-file-p default-directory)))
```

Reply:
Assuming that the behavior was added on Emacs 27.1, shouldn't the condition to use the new behavior exclude Emacs 27.0?

Also, thinking again, is the major version comparison sufficient? `(< 27 emacs-major-version)` will be true for Emacs 28.x, but false for Emacs 27.x, e.g. Emacs 27.1.

I could be wrong, but it seems to me that a correct version check would be one of these previously two commits, which are functionally equivalent as far as I can tell:
- https://github.com/abo-abo/swiper/commit/7314227f38c7a9619957520bf799c0082cb11869
- https://github.com/abo-abo/swiper/commit/7630a885613ac86f7b508792b4ae5372d870069b

Reply:
> Assuming that the behavior was added on Emacs 27.1, shouldn't the condition to use the new behavior exclude Emacs 27.0?

There is no such thing as `27.0`.  Official Emacs releases are versioned as `X.Y`, development versions as `X.Y.Z`.  In both cases, `X` is the `emacs-major-version`, `Y` is the `emacs-minor-version`, and the whole version string is the `emacs-version`.  Development of a new release starts with `X.0.50`, but we only care about `X` anyway.

In other words, `emacs-major-version` is `27` in Emacs `27.0.50`, `27.0.90`, `27.0.91`, and in what will be `27.1`.

Since the optional argument to `executable-find` was added in Emacs 27, we need only check that `emacs-major-version` is greater or equal to `27` before passing the argument.

Commit 7314227f38c7a9619957520bf799c0082cb11869 is mostly correct, but it uses string comparison where a simple arithmetic comparison will do.  It also unnecessarily leaves out pretest versions such as `27.0.90`.

Commit 7630a885613ac86f7b508792b4ae5372d870069b is entirely correct, but it uses more complex string comparison where a simple arithmetic comparison will do.

My suggestion from the very start was to use a simple arithmetic comparison that works for all versions of Emacs 27.

Reply:
Please also pay attention to the order of arguments that I pass to `(< x y)` and `(if a b c)`

Reply:
> @basil-conto: Please also pay attention to the order of arguments that I pass to (< x y) and (if a b c); they are not always in the same order that you seem to be using.

I'm not sure what is meant by this, but please take a look at the current diff to see if the change looks good.

Reply:
> I'm not sure what is meant by this

I meant that, in some of your commits, the order of arguments were the wrong way around, leading e.g. to an off-by-one error.

Reply:
Ah, understood. I thought you meant the current branch diff based on the date you left the comment.